ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,884 @@
+      MessageStoreRecovery recovery, Time time, UUID sessionId, UUID incarnationId) {
+    this.dataDir = new File(dataDir);
+    this.storeId = storeId;
+    this.storeKeyFactory = storeKeyFactory;
+    this.config = config;
+    this.srcMetrics = metrics;
+    tgtMetrics = new StoreMetrics(storeId + METRICS_SUFFIX, metrics.getRegistry());
+    this.srcLog = srcLog;
+    this.diskIOScheduler = diskIOScheduler;
+    this.scheduler = scheduler;
+    this.recovery = recovery;
+    this.time = time;
+    this.sessionId = sessionId;
+    this.incarnationId = incarnationId;
+  }
+
+  /**
+   * If a compaction was in progress during a crash/shutdown, fixes the state so that the store is loaded correctly
+   * and compaction can resume smoothly on a call to {@link #resumeCompaction()}. Expected to be called before the
+   * {@link PersistentIndex} is instantiated in the {@link BlobStore}.
+   * @throws IOException if the {@link CompactionLog} could not be created or if commit or cleanup failed.
+   * @throws StoreException if the commit failed.
+   */
+  void fixStateIfRequired() throws IOException, StoreException {
+    if (CompactionLog.isCompactionInProgress(dataDir.getAbsolutePath(), storeId)) {
+      compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, storeKeyFactory, time);
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case COMMIT:
+          commit(true);
+          compactionLog.markCleanupStart();
+          // fall through to CLEANUP
+        case CLEANUP:
+          cleanup(true);
+          compactionLog.markCycleComplete();
+          break;
+        case COPY:
+          recoveryStartToken = compactionLog.getSafeToken();
+          // fall through to the break.
+        case PREPARE:
+        case DONE:
+          break;
+        default:
+          throw new IllegalStateException("Unrecognized compaction phase: " + phase);
+      }
+      if (compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+        endCompaction();
+      }
+    }
+  }
+
+  /**
+   * Initializes the compactor and sets the {@link PersistentIndex} to copy data from.
+   * @param srcIndex the {@link PersistentIndex} to copy data from.
+   */
+  void initialize(PersistentIndex srcIndex) {
+    this.srcIndex = srcIndex;
+    if (compactionLog == null && srcIndex.hardDeleter != null && srcIndex.hardDeleter.isPaused()) {
+      srcIndex.hardDeleter.resume();
+    }
+    isActive = true;
+  }
+
+  /**
+   * Closes the compactor and waits for {@code waitTimeSecs} for the close to complete.
+   * @param waitTimeSecs the number of seconds to wait for close to complete.
+   * @throws InterruptedException if the wait for close was interrupted.
+   */
+  void close(long waitTimeSecs) throws InterruptedException {
+    isActive = false;
+    if (waitTimeSecs > 0 && !runningLatch.await(waitTimeSecs, TimeUnit.SECONDS)) {
```

Comment:
0 indicates that a wait was never intended in the first place. Not having the first check can result in an error log being printed that says that "compactor did not shutdown in 0 seconds" which seems odd.